### PR TITLE
UCC/UTIL: Components framework

### PR DESCRIPTION
**Pull request by vspetrov**
**What?**
Adds a framework to dynamically load .so components.

**Why?**
Currently it is used to load CL components. Will also be used for TL components and possibly others.

### CODE REVIEW

File: src/utils/ucc_component.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        globfree(&globbuf);
+    }
+    if (n_loaded) {
+        framework->components   = ifaces;
+        framework->n_components = n_loaded;
+    }
+    else {
+        framework->n_components = 0;
+        framework->components   = NULL;
+        return UCC_ERR_NO_MESSAGE;
```
Comment (alex--m) [inline src/utils/ucc_component.c:112]: UCC_ERR_NO_COMPONENTS_FOUND ? this one looks too generic

File: src/utils/ucc_component.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                                           const char *framework_name,
+                                           ucc_component_iface_t **c_iface)
+{
+    char                   iface_struct[128];
+    char                   pattern[128];
+    void                  *handle;
+    ucc_component_iface_t *iface;
+    int                    pos;
+
+    sprintf(pattern, "ucc_%s_", framework_name);
```
Comment (alex--m) [inline src/utils/ucc_component.c:31]: snprintf or assert of strlen(framework_name)

File: src/core/ucc_constructor.c [hunk 3]
```diff
@@ -17,7 +19,7 @@ static int callback(struct dl_phdr_info *info, size_t size, void *data)
     char *component_path;
     if (NULL != (str = strstr(info->dlpi_name, "libucc.so"))) {
         int pos        = (int)(str - info->dlpi_name);
-        component_path = (char *)ucc_malloc(pos + 8);
+        component_path = (char *)ucc_malloc(pos + 8, "component_path");
```
Comment (alex--m) [inline src/core/ucc_constructor.c:22]: #define LIB_UCC_SO ("libucc.so")
#define LIB_UCC_SO_LEN (strlen(LIB_UCC_SO))

File: src/core/ucc_constructor.c [hunk 4]
```diff
@@ -53,6 +55,19 @@ ucc_status_t ucc_constructor(void)
...
         }
     }
+
+    if (!ucc_global_config.component_path) {
+        //TODO
+        //ucc_error, "Failed to get ucc library path. set UCC_COMPONENT_PATH.\n");
+        return UCC_ERR_NO_MESSAGE;
+    }
+    if (UCC_ERR_NO_MESSAGE ==
+        ucc_components_load("cl", &ucc_global_config.cl_framework)) {
```
Comment (alex--m) [inline src/core/ucc_constructor.c:65]: shouldn't "cl" be some kind of parameter? I assume we'll need the same for "tl", "topo" and others...

File: src/core/ucc_constructor.c [hunk 5]
```diff
@@ -53,6 +55,19 @@ ucc_status_t ucc_constructor(void)
             get_default_lib_path();
         }
     }
+
+    if (!ucc_global_config.component_path) {
+        //TODO
+        //ucc_error, "Failed to get ucc library path. set UCC_COMPONENT_PATH.\n");
+        return UCC_ERR_NO_MESSAGE;
+    }
+    if (UCC_ERR_NO_MESSAGE ==
```
Comment (Sergei-Lebedev) [inline src/core/ucc_constructor.c:64]: Why checking only for UCC_ERR_NO_MESSAGE?

File: src/utils/ucc_component.c (hunk 1 above)
Comment (vspetrov) [reply src/utils/ucc_component.c:112]: I've added UCC_ERR_NOT_FOUND - as a generic ucc_status that represents the error. NO_COMPONENTS is too specific for this particular case imho to be added to a list of UCC statuses.

File: src/utils/ucc_component.c (hunk 2 above)
Comment (vspetrov) [reply src/utils/ucc_component.c:31]: reworked this: added UCC_MAX_FRAMEWORK_NAME_LEN and added checks for strlen(framework_name).

File: src/core/ucc_constructor.c (hunk 3 above)
Comment (vspetrov) [reply src/core/ucc_constructor.c:22]: done

File: src/core/ucc_constructor.c (hunk 4 above)
Comment (vspetrov) [reply src/core/ucc_constructor.c:65]: HEre in constructor we are actually loading "CL" components, this is why passing "cl" as parameter to the component framework loader. So, it is a parameter. At some point we will add similar "ucc_components_load" line for "tl" as well.

File: src/utils/ucc_component.c [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    handle = dlopen(so_path, RTLD_LAZY);
+    if (!handle) {
+        //TODO use ucc_error
+        fprintf(stderr,
+                "Failed to load UCC component library: %s. "
+                "Check UCC_COMPONENT_PATH or LD_LIBRARY_PATH\n",
+                so_path);
+        goto error;
+    }
+    iface = (ucc_component_iface_t *)dlsym(handle, iface_struct);
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/utils/ucc_component.c:48]: done

File: src/utils/ucc_component.c [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    sprintf(full_pattern, "%s/ucc_%s_*.so", ucc_global_config.component_path,
+            framework_name);
+    glob(full_pattern, 0, NULL, &globbuf);
+    free(full_pattern);
+    n_loaded          = 0;
+    ifaces_array_size = 0;
+    for (i = 0; i < globbuf.gl_pathc; i++) {
+        if (n_loaded == ifaces_array_size) {
+            ifaces_array_size += 8;
+            ifaces = (ucc_component_iface_t **)ucc_realloc(
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/utils/ucc_component.c:91]: done

File: src/utils/ucc_component.c [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    if (globbuf.gl_pathc > 0) {
+        globfree(&globbuf);
+    }
+    if (n_loaded) {
+        framework->components   = ifaces;
+        framework->n_components = n_loaded;
+    }
+    else {
+        framework->n_components = 0;
+        framework->components   = NULL;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/utils/ucc_component.c:111]: done

File: src/core/ucc_constructor.c (hunk 5 above)
Comment (vspetrov) [reply src/core/ucc_constructor.c:64]: no reason, fixed

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/utils/ucc_component.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+} ucc_component_iface_t;
+
+typedef struct ucc_component_framework {
+    char                   *framework_name;
+    int                     n_components;
+    ucc_component_iface_t **components;
+} ucc_component_framework_t;
+
+/* ucc_components_load searches for all available dynamic components 
+   with the name matching the patter: ucc_<framework_name>_*.so.
```
Comment (manjugv) [inline src/utils/ucc_component.h:25]: Minor: patter -> pattern

File: src/utils/ucc_component.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+typedef struct ucc_component_framework {
+    char                   *framework_name;
+    int                     n_components;
+    ucc_component_iface_t **components;
+} ucc_component_framework_t;
+
+/* ucc_components_load searches for all available dynamic components 
+   with the name matching the patter: ucc_<framework_name>_*.so. 
+   The search is performed in the ucc_global_config.component_path.
+   Each dynamic componet must have a component interface structure defined.
```
Comment (manjugv) [inline src/utils/ucc_component.h:27]: Minor: componet -> component

**Review by manjugv — APPROVED** [review]

File: src/utils/ucc_component.h (hunk 9 above)
Comment (vspetrov) [reply src/utils/ucc_component.h]: fxied

File: src/utils/ucc_component.h (hunk 10 above)
Comment (vspetrov) [reply src/utils/ucc_component.h]: fixed

**Review by alex--m — APPROVED** [review]

**Pull request merged**